### ISSUE TITLE
exp: Queuing experiment causes "Unable to acquire lock" error although no lock file exists at first

### ISSUE THREAD

**Issue by aschuh-hf** [iterative/dvc#8621]
# Bug Report

## Description

I am trying to queue an experiment. There is no `.dvc/tmp/lock` file and no `.dvc/tmp/rwlock.lock` file or any other lock file in the `.dvc/tmp` directory before running `dvc exp run --queue`. During the execution of this command, different lock files are being created (`lock`, `exp_scm_lock`, and multiple `<sha>.lock` files) in `.dvc/tmp/`. Eventually, the command fails with error message:

```
ERROR: Unable to acquire lock. Most likely another DVC process is running or was terminated abruptly. Check the page <https://dvc.org/doc/user-guide/troubleshooting#lock-issue> for other possible reasons and to learn how to resolve this.
```

Note that I am in a larger mono-repo with a local `.dvc` project folder. Git operations generally seem to take a bit longer and I am just wondering if this has to do with the observed error?

Also, I previously had a queue running and stopped it with `dvc queue stop`. This is a long time ago (~1h). I since had uninstalled DVC version 2.31.0 and installed a newer version (tried 2.33.2 and 2.34.0, both installed on CentOS 7 using `yum`).

**Downgrading to DVC version 2.31.0 resolved the issue.**

I had set `core.hardlink_lock = true` in `.dvc/config` previously following the guidance in the error message. This setting made no difference to the observed issue.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
```

**Additional Information (if any):**

**Comment by aschuh-hf** [comment]
The only related release note I could find is the reference to https://github.com/iterative/dvc/pull/8469 for release 2.32.0. This may explain why a version just before that release works fine still.

**Comment by aschuh-hf** [comment]
Might this issue be caused by long running commands issued by the VS Code extension? Observing it again after updating DVC to the latest version 2.35.1.

**Comment by aschuh-hf** [comment]
DVC VS Code Extension output using DVC 2.31.0:

```
[version: 0.5.17, 2022-11-24T18:30:59.208Z, pid: 76555] > dvc plots diff -o .dvc/tmp/plots --split --json - INITIALIZED
[version: 0.5.17, 2022-11-24T18:31:01.758Z, pid: 76257] > dvc exp apply exp-learn-dvf-vxm-loss-img - FAILED with code 255 (25198ms)
[31mERROR[39m: Unable to acquire lock. Most likely another DVC process is running or was terminated abruptly. Check the page <[36mhttps://dvc.org/doc/user-guide/troubleshooting#lock-issue[39m> for other possible reasons and to learn how to resolve this.
[version: 0.5.17, 2022-11-24T18:31:02.136Z, pid: 76555] > dvc plots diff -o .dvc/tmp/plots --split --json - COMPLETED (2961ms)
```

**Comment by aschuh-hf** [comment]
Going to close this issue for now. I recovered from the previous error by removing the `.dvc/tmp/rwlock.lock` file which caused the "checkout" done after applying the workspace changes to fail.

**Comment by shcheklein** [comment]
> which caused the "checkout" done after applying the workspace changes to fail.

could you clarify please?

> I recovered from the previous error by removing the .dvc/tmp/rwlock.lock

doesn't seem like a good solution. Do you understand why it was failing in the first place?

**Comment by aschuh-hf** [comment]
> Do you understand why it was failing in the first place?

No, I don't have a good description as to when this occurs. That's why I though I'll close the issue for now. But indeed there may be something wonky happening making this to occur every now and then. Could also be related to stopping queue workers?

**Comment by aschuh-hf** [comment]
I can confirm this keeps happening. Either `.dvc/tmp/lock` or more commonly `.dvc/tmp/rwlock.lock` is left behind.

Just another piece information in case it is useful, I am running inside a Docker container and the workspace is a mounted volume from the host. I am in the integrated terminal of VS Code and have the DVC extension (v0.5.18) installed and enabled. DVC version currently is 2.35.1 installed with `pip`.

Here's an edited snapshot of my recent bash history where I've marked commands which encountered the "unable to acquire lock" with "ERROR" and the subsequent successful run after manually removing one of the lock files with "OK":
```
 1279  git branch
 1280  git branch -D ...
 1281  dvc exp show
 1282  git branch
 1283  git tab
 1284  git tag
 1285  git checkout -b e1bfe01 ...
 1286  git checkout -b ... e1bfe01
 1287  git status
 1288  git stash -u
 1289  git checkout -b ... e1bfe01
 1290  git rebase -i HEAD~3
 1291  git status
 1292  git branch
 1293  git branch -D ...
 1294  git checkout ...
 1295  dvc exp list --rev ...  # ERROR
 1296  which dvc
 1297  dvc exp list --rev ...  # ERROR
 1298  dvc exp show -n 1  # no failure
 1299  dvc exp list  # ERROR
 1300  ps aux | grep dvc
 1301  rm -f .dvc/tmp/lock
 1302  dvc exp list --rev ...  # OK
 1303  git branch
 1304  git tag
 1305  git branch
 1306  git checkout ...
 1307  git log
 1308  cat metrics.json 
 1309  dvc checkout  # ERROR
 1310  rm -f .dvc/tmp/rwlock.lock 
 1311  dvc checkout  # OK
 1312  less params.yaml 
 1313  git branch
 1314  git branch -m ...
 1315  git tag
 1316  git tag -h
 1317  git tag -m "..." ...
 1318  git tag
 1319  git tag -d ...
 1320  git log
 1321  git tag -l
 1322  git tag -n9
 1323  dvc exp run train --single-item --queue --rev ...  # ERROR
 1324  rm -f .dvc/tmp/rwlock.lock 
 1325  dvc exp run train --single-item --queue --rev ...  # OK
```

**Comment by aschuh-hf** [comment]
I just noticed the `.dvc/tmp/rwlock.lock` file again and compared the PID noted in it to the VS Code DVC Extension console output. It matches the following long running process that was just finishing:
```
[version: 0.5.18, 2022-11-24T19:47:56.534Z, pid: 129714] > /opt/conda/envs/multiseries/bin/python -m dvc exp show --no-fetch -n 3 --json - COMPLETED (45855ms)
```

See also: https://github.com/iterative/dvc/issues/8613#issuecomment-1326725966

**Comment by aschuh-hf** [comment]
And here a scenario which may have caused the `.dvc/tmp/lock` file to remain:

```
$ dvc exp remove 4ed6da6
Removed experiments: 4ed6da6                                          
$ dvc queue remove 4ed6da6
ERROR: '4ed6da6' is not a valid experiment name
$ cat .dvc/tmp/lock 
 24089
$ ps -p 24089 -o command
COMMAND
```

This time, the VS Code DVC Extension console output does not suggest that this PID belonged to one of the commands executed by the extension, but possibly the failed `dvc queue remove` command?

**Comment by karajan1001** [comment]
I implement `rwlock auto solve corrupted lock file` in #8623. It will auto-detect if the process in the lock file still alive. If not it will auto-clean the lock. The lock file problem will be fixed after it is merged.

**Comment by karajan1001** [comment]
> ERROR: Unable to acquire lock. Most likely another DVC process is running or was terminated abruptly. Check the page <https://dvc.org/doc/user-guide/troubleshooting#lock-issue> for other possible reasons and to learn how to resolve this.

This might be solved after #8623 merged. it introduces a lock file double-check by process aliving.

> Also, I previously had a queue running and stopped it with dvc queue stop. This is a long time ago (~1h). I since had uninstalled DVC version 2.31.0 and installed a newer version (tried 2.33.2 and 2.34.0, both installed on CentOS 7 using yum).

Yes, might be introduced by https://github.com/iterative/dvc/pull/8469  need a closer look.

**Comment by karajan1001** [comment]
> And here a scenario which may have caused the `.dvc/tmp/lock` file to remain:
> 
> ```
> $ dvc exp remove 4ed6da6
> Removed experiments: 4ed6da6                                          
> $ dvc queue remove 4ed6da6
> ERROR: '4ed6da6' is not a valid experiment name
> $ cat .dvc/tmp/lock 
>  24089
> $ ps -p 24089 -o command
> COMMAND
> ```
> 
> This time, the VS Code DVC Extension console output does not suggest that this PID belonged to one of the commands executed by the extension, but possibly the failed `dvc queue remove` command?
> 
> I tried then again `dvc queue remove`, but no lock file was produced.
> 
> ```
> $ rm -f .dvc/tmp/lock .dvc/tmp/rwlock.lock 
> $ dvc queue remove 4ed6da6
> ERROR: '4ed6da6' is not a valid experiment name
> ```
> 
> So maybe it was `dvc exp remove` rather which despite successful left the `.dvc/tmp/lock` file behind?!?
> 
> UPDATE: After disabling the VS Code DVC extension, the same steps `dvc exp remove` and `dvc queue remove` did not leave any `lock` file behind (i.e., `dvc exp remove` cleans it up once done). Seems the issue is indeed related to the VS Code DVC extension.

Hi, @aschuh-hf, First thank you for all your reports in recent days. 

here `.dvc/tmp/lock` would not cause the `Unable to acquire lock Error` only the `.dvc/tmp/rwlock` can. And because we are using a granular lock in `.dvc/tmp/rwlock`, only if the contents in `.dvc/tmp/rwlock` have a conflict will cause this error. 

In #8623, we introduce a double-check to check if the process in `.dvc/tmp/rwlock` still exists, it can solve this issue. 

> Also, I previously had a queue running and stopped it with dvc queue stop. This is a long time ago (~1h). I since had uninstalled DVC version 2.31.0 and installed a newer version (tried 2.33.2 and 2.34.0, both installed on CentOS 7 using yum).

This problem has a probability to occur if the experiment running was interrupted. So might not be related to a special version. Could you please elaborate what do you mean here `Downgrading to DVC version 2.31.0 resolved the issue.`
Does it mean that when you get the error occur, and switch back to version `2.31.0` the command will success, or in `2.31.0` the problem will not be triggered?

**Comment by aschuh-hf** [comment]
> In https://github.com/iterative/dvc/pull/8623, we introduce a double-check to check if the process in .dvc/tmp/rwlock still exists, it can solve this issue.

I've seen this issue, but nonetheless I experienced this issue only with DVC versions which include this PR. Could it be that the feature is not working as intended?

> Could you please elaborate what do you mean here Downgrading to DVC version 2.31.0 resolved the issue.
Does it mean that when you get the error occur, and switch back to version 2.31.0 the command will success, or in 2.31.0 the problem will not be triggered?

By downgrading I mean installing an earlier version after I encountered the issue. I still had to manually delete lock files, I think, but with the older DVC version I did then no longer encounter this issue.
